### PR TITLE
Fix Undefined Value Colouring Bug

### PR DESCRIPTION
**Pull request by emmahodcroft**
Should resolve bug #785 , where a) branches without a value get a colour and b) undefined values show up in the legend.

You can see the before examples on [Heroku](https://nextstrain-dev.herokuapp.com/enterovirus/d68/vp1).
I've tested this with a bunch of EV-D68 stuff but it would benefit from someone having a good play with it on something else, esp ensuring filters etc work.

Compare before and after (before always above, after always below):

For categorical without specified colours:
image
image

image
image

For continuous:
image
image

For categorical with specified colors:
image
image
(Tree looks the same)

For clades ('Unassigned' wasn't being detected properly):
image
image

------

As a bonus, managed to get of annoying pie chart lines when filtering so demes go to one colour only:

Before
image

After
image

### CODE REVIEW

File: src/components/map/mapHelpersLatLong.js [hunk 1]
```diff
@@ -73,38 +74,47 @@ const getVisibleNodesPerLocation = (nodes, visibility, geoResolution) => {
...
   });
+
   /* turn counts into arc angles (radians) */
   let startAngle = 0;
   arcs.forEach((a) => {
     a.startAngle = startAngle;
     startAngle += 2*Math.PI*a._count/visibleNodes.length;
     a.endAngle = startAngle;
+    if (a.startAngle === a.endAngle) {
+      // this prevents drawing a 'line' for 'empty' slices
```
Comment (jameshadfield) [inline src/components/map/mapHelpersLatLong.js]: 🥇

File: src/components/map/mapHelpersLatLong.js [hunk 2]
```diff
@@ -445,6 +457,8 @@ UPDATE DEMES & TRANSMISSIONS
 
 const updateDemeDataColAndVis = (demeData, demeIndices, nodes, visibility, geoResolution, nodeColors, pieChart, colorBy, legendValues) => {
   const demeDataCopy = demeData.slice();
+  // get all possible values from tree rather than relying on legend (doesn't include 'undefined', or "", etc)
+  const allValues = Object.keys(countTraitsAcrossTree(nodes, [colorBy], false, false)[colorBy]);
```
Comment (jameshadfield) [inline src/components/map/mapHelpersLatLong.js]: Haven't had a change to test this / read through everything, but I _think_ you'll have to sort this to keep the nice behavior where the arc colours match the order in the legend values.

Comment (emmahodcroft) [reply src/components/map/mapHelpersLatLong.js]: Oh! Hadn't even realised it was doing this. Ok, can do!

**Comment by emmahodcroft** [comment]
This had previously broken colorby Genotype. Now should be fixed...

File: src/components/map/mapHelpersLatLong.js (hunk 2 above)
Comment (emmahodcroft) [reply src/components/map/mapHelpersLatLong.js]: Should be working now - thanks for pointing this out!

**Comment by emmahodcroft** [comment]
Pies should now draw in order of legends, with unknowns at the end.

Found a bug in continuous colouring in `v2` JSONs; should be fixed now.

**Comment by jameshadfield** [comment]
Hey @emmahodcroft -- thanks for the super detailed examples, they were very helpful. I took a good look at this after merging Richard's ordinal color scales and, partly due to the merge conflicts, ended up implementing a slightly different fix (which was super easy since you did all the hard work to identify what was wrong!). Let me talk you through the 3 commits (comments welcome):

---

The main commit was https://github.com/nextstrain/auspice/commit/162824bbed4789a58e1f45930e0dfdc12806fd43. 

Instead of things like `(val === "undefined" || val === undefined || val === false || val === "")` I've  used the `isValueValid` function. This helps keep all the related logic in one place. 

By pruning all "invalid" values from the legend we almost solved all the problems -- the only remaining thing was the arcs, as there were now gaps where a grey arc should be! I fixed this by creating an arc at the end of the list of arcs & assigning tip values which weren't present in the legend to this arc. This lets us avoid extra calls to `countTraitsAcrossTree` :) 

---

https://github.com/nextstrain/auspice/commit/fd961d20891b09e20654ba00abf9ff06b3ec092c is your fix for the arcs with no tips assigned 💯 

---

https://github.com/nextstrain/auspice/commit/50faf2d8689aa4ee55bcb9e0856034f33955c8c6
adds "unassigned" (not case sensitive) to the set of "invalid" values.

**Comment by emmahodcroft** [comment]
Ah ha! I had an idea there might be a better way to do this, which is one reason I was keen for someone else to take a look! This sounds good to me, and more robust :)

**Pull request closed without merging**